### PR TITLE
delve: add linux-riscv64 support

### PR DESCRIPTION
**Pull request by lrzlin**
This patch adds linux-riscv64 support to delve, tested under go 1.22.5/ubuntu 24.04 riscv64 & go 1.21.4/openEuler 24.03 riscv64.

It's worth noticing that due to riscv64 do not support `PTRACE_SINGLESTEP`, just like arm32, so we insert `ebreak` to emulate it. The `__NR_riscv_flush_icache` syscall is used to avoid self modify code's TOCTOU error, but I wasn't able to find any difference with `sys.PtracePokeData`. So it still needs some kinds of validation.

To support CGO, we have to insert `c.ebreak` instead of `ebreak` as breakpoint due to gcc/clang will enable c extension by default, and use `ebreak` as altbreakpoint, but due to a arm64 specific workaround, it can not handle hardcoded breakpoint correctly, so I disabled it on riscv64 platform.

Most of tests were passed, the few failed tests seems to be related with riscv64 go compiler itself, for example, the strange repetition in `TestTraceDirRecursion`, just like this:
```
> goroutine(1): main.A(5, 5)
> goroutine(1): main.A(4, 4)
> goroutine(1): main.A(4, 4)
> goroutine(1): main.A(3, 3)
> goroutine(1): main.A(2, 2)
> goroutine(1): main.A(2, 2)
> goroutine(1): main.A(1, 1)
> goroutine(1): main.A(1, 1)
>> goroutine(1): main.A => (1)
>> goroutine(1): main.A => (2)
>> goroutine(1): main.A => (6)
>> goroutine(1): main.A => (24)
>> goroutine(1): main.A => (120)
120
```
I tried to disable the asyncpreemt, but unfortunately, It doesn't work.

Also, the riscv64 x/arch support hasn't been merged to upstream, so I use `-mod=vendor` here as a workaround.

### CODE REVIEW

File: pkg/proc/native/support_sentinel_linux.go [hunk 1]
```diff
@@ -1,4 +1,4 @@
-//go:build linux && !amd64 && !arm64 && !386 && !(ppc64le && exp.linuxppc64le)
+//go:build linux && !amd64 && !arm64 && !386 && !(ppc64le && exp.linuxppc64le) && !riscv64
```
Comment (aarzilli) [inline pkg/proc/native/support_sentinel_linux.go:1]: If we don't have a build agent for riscv64 this needs to be guarded with a build flag like ppc64le.

File: pkg/proc/native/threads_linux_riscv64.go [hunk 2]
```diff
@@ -0,0 +1,324 @@
...
+	fpregs := riscv64_fpregs.Decode()
+
+	if err != nil {
+		err = fmt.Errorf("could not get floating point registers: %v", err.Error())
+	}
+
+	return fpregs, riscv64_fpregs.Fregs, err
+}
+
+func pokeMemory(pid int, addr uintptr, data []byte) (count int, err error) {
```
Comment (aarzilli) [inline pkg/proc/native/threads_linux_riscv64.go:32]: Why is this needed instead of using the normal way we write to memory?

File: pkg/proc/riscv64_arch.go [hunk 3]
```diff
@@ -0,0 +1,365 @@
...
+func riscv64SwitchStack(it *stackIterator, callFrameRegs *op.DwarfRegisters) bool {
+	if it.frame.Current.Fn == nil {
+		if it.systemstack && it.g != nil && it.top {
+			it.switchToGoroutineStack()
+			return true
+		}
+		return false
+	}
+	switch it.frame.Current.Fn.Name {
+	case "runtime.cgocallback_gofunc", "runtime.cgocallback":
```
Comment (aarzilli) [inline pkg/proc/riscv64_arch.go:151]: This stuff doesn't work or the cgo stacktrace tests would work, delete it.

File: pkg/proc/native/threads_linux_riscv64.go (hunk 2 above)
Comment (lrzlin) [reply pkg/proc/native/threads_linux_riscv64.go:32]: Originally I think we might have a TOCTOU issue when emulate singlestep by insert a ebreak instruction, so I use it for sync the cache to avoid that, but seems that delve will stop every other threads while debugging, if so we don't need it.

File: _scripts/make.go [hunk 4]
```diff
@@ -51,6 +51,8 @@ func NewMakeCommands() *cobra.Command {
 				envflags = append(envflags, "GOOS="+OS)
 			}
 			if len(envflags) > 0 {
+				// Add -mod=vendor to envflags due to riscv arch package has not upstream yet.
+				envflags = append(envflags, "-mod=vendor")
```
[the remark this answers is not in the document]
Comment (lrzlin) [reply _scripts/make.go:55]: This is just a workaround, we will remove that once the arch package support was upstream.

File: pkg/proc/native/support_sentinel_linux.go (hunk 1 above)
Comment (lrzlin) [reply pkg/proc/native/support_sentinel_linux.go]: Added.

File: pkg/proc/riscv64_arch.go (hunk 3 above)
Comment (lrzlin) [reply pkg/proc/riscv64_arch.go:151]: Deleted.

File: pkg/proc/native/threads_linux_riscv64.go (hunk 2 above)
Comment (lrzlin) [reply pkg/proc/native/threads_linux_riscv64.go]: I just runs the test without this custom `pokeMemory`, and it's works well, so we can safely restore using `sys.PtracePokeData`.

**Comment by lrzlin** [comment]
@aarzilli I was working on cgostacktrace for riscv64, it seems that both `arm64cgocallSPOffsetSaveSlot` and ` prevG0schedSPOffsetSaveSlot`'s value is wrong, the second one should be `24`, but I couldn't find the first one in `asmcgocall`, where should I find it? 

Moreover, ppc64le do follows the numbers found in `asm_ppc64x.s` but it still couldn't use cgostacktrace, So did this feature needs golang side support, or it's just because the logic of `riscv64SwitchStack` is simply wrong? Thanks.

**Comment by aarzilli** [comment]
I don't remember how the values for arm64 were derived and, as far as I know, the ppc64 implementation never worked so it's possible that the logic is wrong.

File: _scripts/make.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (lrzlin) [reply _scripts/make.go]: Removed this workaround, riscv64 support is already merged into x/arch, see https://github.com/golang/arch/commit/7874f23b9c060aa1d0aaa13b7352d59335c30184

**Comment by lrzlin** [comment]
@aarzilli Hi, riscv64 x/arch support is already merged in upstream, you could find that in https://github.com/golang/arch/commit/7874f23b9c060aa1d0aaa13b7352d59335c30184, I have also updated the code, could you please give this patch another look? Thanks a lot.

File: pkg/proc/linutil/regs_riscv64_arch.go [hunk 5]
```diff
@@ -0,0 +1,386 @@
...
+	"encoding/binary"
+	"fmt"
+
+	"github.com/go-delve/delve/pkg/dwarf/op"
+	"github.com/go-delve/delve/pkg/dwarf/regnum"
+	"github.com/go-delve/delve/pkg/proc"
+	"golang.org/x/arch/riscv64/riscv64asm"
+)
+
+// Regs is wrapper for sys.PtraceRegs
```
Comment (aarzilli) [inline pkg/proc/linutil/regs_riscv64_arch.go:13]: The comment needs to start with "RISCV64Registers is..."

File: pkg/proc/linutil/regs_riscv64_arch.go [hunk 6]
```diff
@@ -0,0 +1,386 @@
...
+		reg.FillBytes()
+		copy(r.Fpregset[8*i:], reg.Bytes)
+		return true, nil
+
+	default:
+		return false, fmt.Errorf("changing register %d not implemented", regNum)
+	}
+}
+
+// Refer to the definition of struct __riscv_d_ext_state in the kernel ptrace.h
```
Comment (aarzilli) [inline pkg/proc/linutil/regs_riscv64_arch.go:360]: The comment needs to start with "RISCV64PtraceFpRegs is..."

File: pkg/proc/core/linux_core.go [hunk 7]
```diff
@@ -333,6 +367,15 @@ func readNote(r io.ReadSeeker, machineType elf.Machine) (*note, error) {
 			}
 			note.Desc = fpregs
 		}
+
+		if machineType == _EM_RISCV {
```
Comment (aarzilli) [inline pkg/proc/core/linux_core.go:371]: else if or convert to a switch.

File: pkg/proc/native/proc.go [hunk 8]
```diff
@@ -372,7 +372,7 @@ func (dbp *nativeProcess) initialize(path string, debugInfoDirs []string) (*proc
 		//	  with gdb once AsyncPreempt was enabled. While implementing the port,
 		//	  few tests failed while it was enabled, but cannot be warrantied that
 		//	  disabling it fixed the issues.
-		DisableAsyncPreempt: runtime.GOOS == "windows" || (runtime.GOOS == "linux" && runtime.GOARCH == "arm64") || (runtime.GOOS == "linux" && runtime.GOARCH == "ppc64le"),
+		DisableAsyncPreempt: runtime.GOOS == "windows" || (runtime.GOOS == "linux" && runtime.GOARCH == "arm64") || (runtime.GOOS == "linux" && runtime.GOARCH == "ppc64le") || (runtime.GOOS == "linux" && runtime.GOARCH == "riscv64"),
```
Comment (aarzilli) [inline pkg/proc/native/proc.go:375]: There is a comment above explaining why we disable async preempt, if it really needs to be disable on riscv64 it should be updated to describe why.

File: pkg/proc/native/support_sentinel_linux.go (hunk 1 above)
Comment (aarzilli) [reply pkg/proc/native/support_sentinel_linux.go:1]: That could be done. The instructions to add a teamcity agent are here: https://www.jetbrains.com/help/teamcity/install-teamcity-agent.html.

File: pkg/proc/riscv64_arch.go (hunk 3 above)
Comment (aarzilli) [reply pkg/proc/riscv64_arch.go:151]: I meant all of it: runtime.asmcgocall and crosscall2 are the same thing as is the switch at the bottom of this function.

File: pkg/proc/native/proc.go (hunk 8 above)
Comment (lrzlin) [reply pkg/proc/native/proc.go:375]: Some tests failed like the TestTraceDirRecursion, the symptom which seems to cause by async preempt but disable it doesn't work just like I said, so I'll enable it, until we find the exact reason.

> Most of tests were passed, the few failed tests seems to be related with riscv64 go compiler itself, for example, the strange repetition in `TestTraceDirRecursion`, just like this:
> 
> ```
> > goroutine(1): main.A(5, 5)
> > goroutine(1): main.A(4, 4)
> > goroutine(1): main.A(4, 4)
> > goroutine(1): main.A(3, 3)
> > goroutine(1): main.A(2, 2)
> > goroutine(1): main.A(2, 2)
> > goroutine(1): main.A(1, 1)
> > goroutine(1): main.A(1, 1)
> >> goroutine(1): main.A => (1)
> >> goroutine(1): main.A => (2)
> >> goroutine(1): main.A => (6)
> >> goroutine(1): main.A => (24)
> >> goroutine(1): main.A => (120)
> 120
> ```
> 
> I tried to disable the asyncpreemt, but unfortunately, It doesn't work.

File: pkg/proc/linutil/regs_riscv64_arch.go (hunk 6 above)
Comment (lrzlin) [reply pkg/proc/linutil/regs_riscv64_arch.go:360]: Fixed.

File: pkg/proc/core/linux_core.go (hunk 7 above)
Comment (lrzlin) [reply pkg/proc/core/linux_core.go:371]: Fixed.

File: pkg/proc/linutil/regs_riscv64_arch.go (hunk 5 above)
Comment (lrzlin) [reply pkg/proc/linutil/regs_riscv64_arch.go:13]: Fixed.

File: pkg/proc/riscv64_arch.go (hunk 3 above)
Comment (lrzlin) [reply pkg/proc/riscv64_arch.go]: `runtime.asmcgocall` and `crosscall2` removed.

File: pkg/proc/native/support_sentinel_linux.go (hunk 1 above)
Comment (lrzlin) [reply pkg/proc/native/support_sentinel_linux.go:1]: Thanks, we could provide riscv hardware for this, then how could we add this in delve's teamcity just like ppc64le? Also, does this ci machine need to be controled by delve dev team directly or we could host it in a third-party organization?

**Comment by lrzlin** [comment]
@aarzilli Hi, the CI machine is ready and I've modified the `.teamcity/settings.kts` file. Also, the teamcity agent is running on this machine, I configured it as `serverUrl=https://delve.teamcity.com/` and `name=Linux-openEuler24.09-riscv64` directly in `conf/buildAgent.properties`, and the log says it needs to be authorized by administrator, is that okay?
```
[2024-10-09 01:22:03,525]   INFO - buildServer.AGENT.registration - Registering on server via URL "https://delve.teamcity.com": AgentDetails{Name='Linux-openEuler24.09-riscv64', AgentId=null, BuildId=null, AgentOwnAddress='null', AlternativeAddresses=[], Port=9090, Version='160695', PluginsVersion='160695-md5-38792efbad1c0fa39ca4698fc8452484', AvailableRunners=[Ant, cargo-deploy-runner, csharpScript, DockerCommand, dotcover, dotnet, dotnet-tools-dupfinder, dotnet-tools-inspectcode, Duplicator, ftp-deploy-runner, gradle-runner, Inspection, jetbrains_powershell, JPS, kotlinScript, Maven2, nodejs-runner, nunit-console, octopus.generic, octopus.metadata, python-runner, Qodana, rake-runner, SBT, simpleRunner, smb-deploy-runner, ssh-deploy-runner, ssh-exec-runner, unity, unreal-engine], AvailableVcs=[tfs, jetbrains.git, mercurial, svn, perforce], AuthorizationToken='', PingCode='zP69uz6e1BIhS305wqaspLKxwexNPChz'}
[2024-10-09 01:22:07,082]   INFO - buildServer.AGENT.registration - Server supports the following communication protocols: [polling]
[2024-10-09 01:22:07,084]   INFO - buildServer.AGENT.registration - Trying to register on server using 'polling' protocol.
[2024-10-09 01:22:08,414]   INFO - ldServer.AGENT.PollingProtocol - Start polling server for commands
[2024-10-09 01:22:08,417]   INFO - buildServer.AGENT.registration - Registered on server with id 11420 and authorization token '3fe84046c543a557327f2ca421ae6763'
[2024-10-09 01:22:08,418]   INFO - buildServer.AGENT.registration - If this is the first time this agent registered on the server make sure it is authorized by administrator in the server web UI.
[2024-10-09 01:22:10,463]   INFO -    jetbrains.buildServer.AGENT - Updating agent parameters on the server: AgentDetails{Name='Linux-openEuler24.09-riscv64', AgentId=11420, BuildId=null, AgentOwnAddress='null', AlternativeAddresses=[], Port=9090, Version='160695', PluginsVersion='160695-md5-38792efbad1c0fa39ca4698fc8452484', AvailableRunners=[Ant, cargo-deploy-runner, csharpScript, DockerCommand, dotcover, dotnet, dotnet-tools-dupfinder, dotnet-tools-inspectcode, Duplicator, ftp-deploy-runner, gradle-runner, Inspection, jetbrains_powershell, JPS, kotlinScript, Maven2, nodejs-runner, nunit-console, octopus.generic, octopus.metadata, python-runner, Qodana, rake-runner, SBT, simpleRunner, smb-deploy-runner, ssh-deploy-runner, ssh-exec-runner, unity, unreal-engine], AvailableVcs=[tfs, jetbrains.git, mercurial, svn, perforce], AuthorizationToken='3fe84046c543a557327f2ca421ae6763', PingCode='zP69uz6e1BIhS305wqaspLKxwexNPChz'}
```
BTW, this machine runs openEuler 24.09, is it has to be Ubuntu or other distro is just fine? Thanks.

File: pkg/dwarf/regnum/riscv64.go [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+func RISCV64ToName(num uint64) string {
+	switch {
+	case num <= RISCV64_X31:
+		return fmt.Sprintf("X%d", num)
+
+	case num >= RISCV64_F0 && num <= RISCV64_F31:
+		return fmt.Sprintf("F%d", num)
+
+	case num == RISCV64_PC:
+		return fmt.Sprintf("PC")
```
Comment (aarzilli) [inline pkg/dwarf/regnum/riscv64.go:69]: just `return "PC"` here

File: pkg/dwarf/regnum/riscv64.go [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+func RISCV64MaxRegNum() uint64 {
+	return _RISC64_MaxRegNum
+}
+
+var RISCV64NameToDwarf = func() map[string]int {
+	r := make(map[string]int)
+	for i := 0; i <= 31; i++ {
+		r[fmt.Sprintf("x%d", i)] = RISCV64_X0 + i
+	}
+	r[fmt.Sprintf("pc")] = RISCV64_PC
```
Comment (aarzilli) [inline pkg/dwarf/regnum/riscv64.go:85]: `r["pc"] = RISCV64_PC`

File: pkg/proc/bininfo.go [hunk 11]
```diff
@@ -1803,6 +1806,9 @@ func (bi *BinaryInfo) setGStructOffsetElf(image *Image, exe *elf.File, wg *sync.
 	case elf.EM_PPC64:
 		_ = getSymbol(image, bi.logger, exe, "runtime.tls_g")
 
+	case elf.EM_RISCV:
```
Comment (aarzilli) [inline pkg/proc/bininfo.go:1809]: change the case above to be `case elf.EM_PPC64, elf.EM_RISCV:`

File: pkg/proc/riscv64_disasm.go [hunk 12]
```diff
@@ -0,0 +1,189 @@
...
+import (
+	"fmt"
+
+	"github.com/go-delve/delve/pkg/dwarf/op"
+	"github.com/go-delve/delve/pkg/dwarf/regnum"
+	"golang.org/x/arch/riscv64/riscv64asm"
+)
+
+var (
+	errType = fmt.Errorf("unknown type")
```
Comment (aarzilli) [inline pkg/proc/riscv64_disasm.go:12]: It doesn't look like this is being used anywhere.

**Review by aarzilli — COMMENTED** [review]
The agent was approved.

PS. note that the changes to the teamcity configuration will not take effect until we merge this PR.

File: pkg/proc/riscv64_disasm.go (hunk 12 above)
Comment (lrzlin) [reply pkg/proc/riscv64_disasm.go]: Deleted.

File: pkg/proc/bininfo.go (hunk 11 above)
Comment (lrzlin) [reply pkg/proc/bininfo.go]: Done.

File: pkg/dwarf/regnum/riscv64.go (hunk 10 above)
Comment (lrzlin) [reply pkg/dwarf/regnum/riscv64.go]: Done.

File: pkg/dwarf/regnum/riscv64.go (hunk 9 above)
Comment (lrzlin) [reply pkg/dwarf/regnum/riscv64.go]: Done.

**Comment by lrzlin** [comment]
> There are merge conflicts with go.mod and vendor/modules.txt, aside from that I think everything is fine.

Resolved, now everything should be okay.

**Review by aarzilli — APPROVED** [review]
LGTM

**Review by derekparker — APPROVED** [review]
LGTM

**Comment by derekparker** [comment]
Thanks for the contribution!
